### PR TITLE
update k8s & related libraries to 1.28

### PR DESCRIPTION
**Pull request by akalenyu**
**What this PR does / why we need it**:
There are a few non trivial changes laid out in the commit descriptions

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
k8s 1.28 dependencies
```

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -170,36 +169,36 @@ replace (
 	github.com/openshift/library-go => github.com/mhenriks/library-go v0.0.0-20230310153733-63d38b55bd5a
```
Comment (mhenriks) [inline go.mod:169]: `openshift/api` (looks like `release-4.15` is the appropriate branch for k8s 1.28) and `openshift/library-go` should probably be updated as well.  library-go is tricky as this was never merged:  https://github.com/openshift/library-go/pull/1486

This is obviously pretty annoying.  I guess I'll update my fork again for now

**Review by mhenriks — COMMENTED** [review]
Should we do 1.29?  That is what openshift is using for 4.16

**Comment by akalenyu** [comment]
> Should we do 1.29? That is what openshift is using for 4.16

Yeah, I plan to do that as well but we'll need a bump to go 1.21 first
Anyway, this PR is too big as it is, it would be nice to do that in follow up

**Comment by mhenriks** [comment]
> > Should we do 1.29? That is what openshift is using for 4.16
> 
> Yeah, I plan to do that as well but we'll need a bump to go 1.21 first Anyway, this PR is too big as it is, it would be nice to do that in follow up

Why can't we go straight to 1.29?

**Comment by akalenyu** [comment]
> > > Should we do 1.29? That is what openshift is using for 4.16
> > 
> > 
> > Yeah, I plan to do that as well but we'll need a bump to go 1.21 first Anyway, this PR is too big as it is, it would be nice to do that in follow up
> 
> Why can't we go straight to 1.29?

Because we are still on go 1.20

File: go.mod (hunk 1 above)
Comment (akalenyu) [reply go.mod:169]: Let me know when that is done and I can update both

Comment (mhenriks) [reply go.mod:169]: Looks like last week the change was made on master and backported to release-4.16 in another PR (https://github.com/openshift/library-go/commit/434fa21f5e457d51f436c4b3804c7803db1da42d)

So if you want to go straight to 1.29 you can use release-4.16 branch in library-go
If you insist on doing 1.28 first (sounds like more work to me) then you can use `github.com/mhenriks/library-go export-certrotation-4.15`

File: cmd/cdi-operator/operator.go [hunk 2]
```diff
@@ -84,7 +85,11 @@ func main() {
 	}
 
 	managerOpts := manager.Options{
-		Namespace:                  namespace,
+		Cache: cache.Options{
+			DefaultNamespaces: map[string]cache.Config{
+				namespace: {},
```
Comment (alromeros) [inline cmd/cdi-operator/operator.go:90]: No need to include `namespace` anywhere?

File: cmd/cdi-controller/controller.go [hunk 3]
```diff
@@ -372,24 +385,53 @@ func getTokenPrivateKey() *rsa.PrivateKey {
...
-func getNewManagerCache(cdiNamespace string) cache.NewCacheFunc {
+func getCacheOptions(apiClient client.Client, cdiNamespace string) cache.Options {
 	namespaceSelector := fields.Set{"metadata.namespace": cdiNamespace}.AsSelector()
-	return cache.BuilderWithOptions(
-		cache.Options{
-			SelectorsByObject: cache.SelectorsByObject{
-				&networkingv1.Ingress{}: {
-					Field: namespaceSelector,
-				},
-				&routev1.Route{}: {
```
Comment (alromeros) [inline cmd/cdi-controller/controller.go:383]: Just making sure, we don't want route anymore?

**Review by alromeros — COMMENTED** [review]
Awesome, thanks for dividing it into commits. Just a couple of questions.

File: cmd/cdi-operator/operator.go (hunk 2 above)
Comment (akalenyu) [reply cmd/cdi-operator/operator.go:90]: wdym? it's included `namespace: {},`

File: cmd/cdi-controller/controller.go (hunk 3 above)
Comment (akalenyu) [reply cmd/cdi-controller/controller.go:383]: We conditionally add `&routev1.Route{}` if running on openshift

Comment (akalenyu) [reply cmd/cdi-controller/controller.go:383]: here's some context - https://github.com/kubernetes-sigs/controller-runtime/issues/2456

File: cmd/cdi-operator/operator.go (hunk 2 above)
Comment (alromeros) [reply cmd/cdi-operator/operator.go:90]: I mean the namespace we get above (`namespace := util.GetNamespace()`). Just wondering if we don't have to assign it somewhere as we did before.

Comment (akalenyu) [reply cmd/cdi-operator/operator.go:90]: it's included as the key in `map[string]cache.Config`

File: go.mod (hunk 1 above)
Comment (akalenyu) [reply go.mod:169]: We can't - https://github.com/openshift/library-go/blob/master/go.mod#L3, same for ctrl runtime

Comment (akalenyu) [reply go.mod:169]: Do you mean a few separate PRs to bump go/update builder image, right?

Comment (akalenyu) [reply go.mod:169]: Anyway, this is huge as it is, if more (1.28->1.29) adaptations are needed this PR will just be all over the place
so I prefer to take the 1.28->1.29 stuff out

**Comment by alromeros** [comment]
I agree this PR is already too large, I'm ok with doing the 1.28->1.29 stuff in a different one
/lgtm

**Pull request merged**